connectors/src/connectors/slack/lib/workspace_rate_limit.ts

```diff
@@ -0,0 +1,155 @@
+import type { ModelId } from "@dust-tt/types";
+import { cacheWithRedis, DustAPI } from "@dust-tt/types";
+
+import { getSlackClient } from "@connectors/connectors/slack/lib/slack_client";
+import { dataSourceConfigFromConnector } from "@connectors/lib/api/data_source_config";
+import {
+  addEntityToRateLimiter,
+  isEntityRateLimited,
+} from "@connectors/lib/entity_rate_limiter";
+import type { Connector } from "@connectors/lib/models";
+import { redisClient } from "@connectors/lib/redis";
+import logger from "@connectors/logger/logger";
+
+export async function safeRedisClient<T>(
+  fn: (client: Awaited<ReturnType<typeof redisClient>>) => PromiseLike<T>
+): Promise<T> {
+  const client = await redisClient();
+  try {
+    return await fn(client);
+  } finally {
+    await client.quit();
+  }
+}
+
+async function getActiveMembersCount(connector: Connector): Promise<number> {
+  const ds = dataSourceConfigFromConnector(connector);
+
+  // The number of active members in the workspace.
+  const dustAPI = new DustAPI(
+    {
+      apiKey: ds.workspaceAPIKey,
+      workspaceId: ds.workspaceId,
+    },
+    logger,
+    { useLocalInDev: true }
+  );
+
+  const activeMembersRes = await dustAPI.getActiveMembersCountInWorkspace();
+  if (activeMembersRes.isErr()) {
+    logger.error("Error getting all members in workspace.", {
+      error: activeMembersRes.error,
+    });
+
+    throw new Error("Error getting all members in workspace.");
+  }
+
+  return activeMembersRes.value;
+}
+
+function makeSlackRateLimiterForConnectorKey(connectorId: ModelId): string {
+  return `slack-rate-limiter-connector-${connectorId}`;
+}
+
+export const getActiveMembersCountMemoized = cacheWithRedis(
+  getActiveMembersCount,
+  (connector: Connector) => {
+    return `active-members-connector-${connector.id}`;
+  },
+  // Caches data for 15 minutes to limit frequent API calls.
+  // Note: Updates (e.g., new seats added by an admin) may take up to 15 minutes to be reflected.
+  15 * 10 * 1000
+);
+
+async function postWorkspaceRateLimitedMessage(
+  connector: Connector,
+  {
+    slackChannel,
+    slackMessageTs,
+    maxAllowed,
+  }: { slackChannel: string; slackMessageTs: string; maxAllowed: number }
+) {
+  const slackClient = await getSlackClient(connector.id);
+  await slackClient.chat.postMessage({
+    channel: slackChannel,
+    blocks: [
+      {
+        type: "section",
+        text: {
+          type: "mrkdwn",
+          text: `Your Slack integration is limited to the number of active members in your workspace, which is currently ${maxAllowed} users. To allow more users to access the app, please increase your active member count.`,
```

Comment:
Deferring wording review to @gabhubert

Reply:
As an example "Increase your active member count" is not super clear? It also talks to the "admin" but the person there is the user.

Maybe it should be something more along the lines of "sign-in on Dust" ?

Reply:
Oh yeah sign-in + auto-join domain = 😁.